File: Languages/python/algorithms-dsa/leetcode/monotonic-stack/largest_rectangle_histogram.py
"""
LeetCode #84 - Largest Rectangle in Histogram
Topic: Monotonic Stack
Difficulty: Hard

Given an array of bar heights, find the area of the largest rectangle that
can be formed in the histogram.

Approach: Maintain an increasing monotonic stack of indices. When a bar is
shorter than the stack top, pop and calculate area using the popped bar as
the smallest bar. Width extends from the new stack top to current index.

Time Complexity: O(n) - each bar pushed/popped at most once
Space Complexity: O(n)
"""


def largest_rectangle_area(heights: list[int]) -> int:
    stack: list[int] = []  # increasing stack of indices
    max_area = 0
    n = len(heights)

    for i in range(n + 1):
        # Use 0 as sentinel at the end to flush the stack
        h = heights[i] if i < n else 0

        while stack and h < heights[stack[-1]]:
            height = heights[stack.pop()]
            width = i if not stack else i - stack[-1] - 1
            max_area = max(max_area, height * width)

        stack.append(i)

    return max_area


if __name__ == "__main__":
    assert largest_rectangle_area([2, 1, 5, 6, 2, 3]) == 10
    assert largest_rectangle_area([2, 4]) == 4
    assert largest_rectangle_area([1]) == 1
    assert largest_rectangle_area([1, 1, 1, 1]) == 4
    assert largest_rectangle_area([6, 2, 5, 4, 5, 1, 6]) == 12
    print("All tests passed!")

File: Languages/go/algorithms-dsa/leetcode/monotonic-stack/largest_rectangle_histogram.go
package main

import "fmt"

/*
  LC 84 - Largest Rectangle in Histogram
  Topic: Monotonic Stack
  Difficulty: Hard
  Time: O(n) | Space: O(n)
*/

func largestRectangleArea(heights []int) int {
	stack := []int{}
	maxArea := 0
	for i := 0; i <= len(heights); i++ {
		h := 0
		if i < len(heights) {
			h = heights[i]
		}
		for len(stack) > 0 && h < heights[stack[len(stack)-1]] {
			height := heights[stack[len(stack)-1]]
			stack = stack[:len(stack)-1]
			width := i
			if len(stack) > 0 {
				width = i - stack[len(stack)-1] - 1
			}
			area := height * width
			if area > maxArea {
				maxArea = area
			}
		}
		stack = append(stack, i)
	}
	return maxArea
}

func main() {
	res := largestRectangleArea([]int{2, 1, 5, 6, 2, 3})
	if res == 10 {
		fmt.Println("PASS: largestRectangleArea")
	} else {
		fmt.Println("FAIL: largestRectangleArea, got", res)
	}

	res2 := largestRectangleArea([]int{2, 4})
	if res2 == 4 {
		fmt.Println("PASS: largestRectangleArea case 2")
	} else {
		fmt.Println("FAIL: largestRectangleArea case 2, got", res2)
	}
}
